src/main/java/nextstep/subway/line/application/LineService.java

```diff
@@ -1,23 +1,57 @@
 package nextstep.subway.line.application;
 
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.persistence.EntityNotFoundException;
+
+import org.springframework.dao.EmptyResultDataAccessException;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+
 import nextstep.subway.line.domain.Line;
 import nextstep.subway.line.domain.LineRepository;
 import nextstep.subway.line.dto.LineRequest;
 import nextstep.subway.line.dto.LineResponse;
-import org.springframework.stereotype.Service;
-import org.springframework.transaction.annotation.Transactional;
 
 @Service
 @Transactional
 public class LineService {
-    private LineRepository lineRepository;
 
-    public LineService(LineRepository lineRepository) {
-        this.lineRepository = lineRepository;
-    }
+	private final LineRepository lineRepository;
+
+	public LineService(LineRepository lineRepository) {
+		this.lineRepository = lineRepository;
+	}
+
+	public LineResponse saveLine(LineRequest request) {
+		return LineResponse.of(lineRepository.save(request.toLine()));
+	}
+
+	public List<LineResponse> getLines() {
+		return lineRepository.findAll().stream()
+			.map(LineResponse::of)
+			.collect(Collectors.toList());
+	}
+
+	public LineResponse getLine(Long id) {
+		return LineResponse.of(findLine(id));
+	}
+
+	public void updateLine(Long id, LineRequest request) {
+		final Line line = findLine(id);
+		line.update(request.toLine());
+	}
+
+	public void deleteLine(Long id) {
+		if (lineRepository.existsById(id)) {
+			lineRepository.deleteById(id);
+		}
```

Comment:
select 문이 2번 수행되는 것을 의도한 것이 맞습니다.
이유는 `repo.deleteById()` 호출시 삭제할 데이터가 없는 경우 `EmptyResultDataAccessException` 예외가 발생하는데요. 제가 이 예외를 service 레벨에서 try-catch 처리하고 `deleteLine_notFound` 테스트를 수행하니 controller 레벨에서 아래 예외가 발생하더군요.
```
org.springframework.transaction.UnexpectedRollbackException: 
Transaction silently rolled back because it has been marked as rollback-only
```
[자료](https://techblog.woowahan.com/2606/)를 찾아보니 아래와 같은 이유가 있더라구요.

> 참여 중인 트랜잭션이 실패하면 기본정책이 전역롤백이다.

그래서 고민한게..

1. (자료에서 언급된) `PROPAGATION_NESTED` 설정 => `DataSourceTransactionManager`를 직접 다뤄야하는데 오버하는 것 같아서 fail.. (그렇다고 `globalRollbackOnParticipationFailure` 속성을 건드리는 건 더 아닌 것 같고..)
2. `RestControllerAdvice`에서 `EmptyResultDataAccessException` 핸들링 => `404`를 내려주는게 일반적인 것 같지만 삭제 요청시 404 를 내려주는게 일반적인지 의문(애초에 요구사항에도 없는 스펙인데다 ~그냥 궁금해서 만들어본 테스트인데 이정도로 파보게 만들어서 당황x100~)
3. 단순무식하지만 정석대로(??) 간다

이런 흐름으로 가게 되서 의도하게 되었습니다.
2.에서 404를 내려줘도 step1 요구사항을 위반하는 건 아니지만 개인적으로 어색한 것 같아 3.으로 처리했습니다.
정하기 나름인 것 같기도 한데.. 전자의 방법이 더 적절할까요? 아니면 더 나은 방법이 있을까요..?

Reply:
요구사항에서 예시 응답이 204로 되어 있었기에 그렇게 고려해주셨던 거였군요!

해당 요구사항은 정상적인 케이스의 응답만을 정의한 것이라 볼 수 있습니다. delete할 데이터가 없는 상황에 대한 응답은 404면 적절합니다. endpoint를 찾을 수 없어서 발생한 에러인지, 데이터를 찾을 수 없어서 난 에러인지 정도만 body에 명시해주면 될 것이라 생각합니다.

Reply:
저는 예외처리를 데이터 삭제할 내용이 없을 시 발생한 에러에 대한 익셉션 핸들러라 생각하고 말씀드린 것이었는데, delete할 데이터가 없는 상황에서도 204 응답을 내려주기 위한 예외처리를 생각하고 계셨던 것이군요!

delete할 내용이 없으면 4xx대 응답을 내려주는 것이 정석입니다. 다만 view의 요구사항에서 꼭 2xx대 코드로 내려달라고 하는 요구사항이 있으면 어쩔 수 없겠죠...

Reply:
delete할 내용이 없을 시 404 내려주도록 변경햇고 `ErrorResponse` 클래스 추가해서 에러 메시지가 내려가도록 처리햇습니다.
명확한 답변 감사합니다 👍